proposals/0002.md

```diff
@@ -0,0 +1,270 @@
+* Title: Structured error handling in Rust
+* Champion: @CodeSandwich
+* Conversation (optional): [thread on radicle.community](https://radicle.community/t/proposal-structured-error-handling-in-rust/179)
+* References:
+  * [The previous error handling proposal](https://github.com/radicle-dev/radicle-decisions/blob/master/proposals/0001.md)
+  * [`std::error::Error`](https://doc.rust-lang.org/std/error/trait.Error.html)
+  * [A discussion about the good design of errors](https://users.rust-lang.org/t/do-people-not-care-about-printable-error-chains-a-k-a-how-to-nicely-implement-display-for-an-error/35362)
+
+# Summary/Motivation/Introduction
+We need good error handling to make our lives easier.
+We've already agreed to use `thiserror` and `anyhow`, which are fantastic tools, but we're still
+lacking a common design language.
+There are many ways to create an error and some of them don't play well with others.
+Without common conventions handling errors is problematic too.
+
+This is a guide to structuring errors well.
+
+# Explanation
+
+## The great `Error` unification
+
+The ultimate goal is to make your errors as easy to integrate and handle as possible.
+The canonical way of achieving this is by implementing the `std::error::Error` trait.
+It's universally used in the Rust ecosystem and provides a common interface for working with
+different errors coming from numerous crates.
+It's not possible to enforce the exact behavior of an implementation of the `Error` trait, but
+there's a widely used convention making it reasonably uniform and predictable.
+
+## The correct behavior of an `Error` implementation
+
+`Error` provides 2 core functionalities:
+
+- it provides the `source` method, which may or may not return a reference to another `Error`
```

Comment:
it'd be good to explain that that "another `Error`" would be a wrapper error.

Reply:
At this point it doesn't have to be. This is a description of the raw interface and it will accept any behavior that will succeed to compile. If you want to return `&self`, or randomize your output, you can do it! What exactly **should** be returned is explained in the following sections.

Reply:
Fair :+1: